Step2 - 문자열 사칙 연산 계산기

maenguin opened the pull request:
문자열 사칙 연산 계산기 구현 PR 입니다.
Main.kt에서 실행 가능합니다.

다음과 같이 구성하였습니다.

Calculator : 전체적인 계산 작업을 오케스트레이션
ㄴ Validator : 연산식 검증
ㄴ Tokenizer : 연산식에서 의미있는 문자만 토큰화
ㄴ Parser : 토큰화된 문자를 구체적인 타입(연산자, 피연산자)로 변환
ㄴ Strategy : 연산 방법을 결정

ExpressionItem : 연산식에 올 수 있는 요소를 하나로 묶는 인터페이스
ㄴ Operand : 피연산자로서 실제 연산을 수행
ㄴ Operator : 연산자로서 문자 연산자를 매핑하고 구체적인 타입으로 매핑하고 연산 메시지를 받는 역할
--ㄴ Plus
--ㄴ Minus
--ㄴ Multiply
--ㄴ Divide

maenguin committed:
kotest StringSpec 학습 테스트 추가

maenguin committed:
step2 : 계산기 프로토타이핑

maenguin committed:
step2 : Junit5 테스트가 동작하지 않아서 kotest-runner-junit5 버전 5.3.2로 원복

maenguin committed:
step2 : 연산식을 검증하는 ExpressionValidator 추가

maenguin committed:
step2 : 문자 연산식을 토큰화하는 ExpressionTokenizer 추가

maenguin committed:
step2 : 토큰화된 문자 연산식을 타입 매핑하는 ExpressionParser 추가

maenguin committed:
step2 : 연산자 피연산자 타입 추가

maenguin committed:
step2 : 전체적인 연산 방법을 결정하는 CalculatorStrategy 추가

maenguin committed:
step2 : 전체적인 계산 작업을 오케스트레이션 하는 Calculator 클래스 추가

maenguin committed:
step2 : 패키지 구조 변경

maenguin committed:
step2 : CalculatorStrategy 테스트 추가

pci2676 commented on src/main/kotlin/step2/ExpressionValidator.kt:
```diff
@@ -0,0 +1,19 @@
+ */
+class ExpressionValidator {
+
+    private val patternThatAllowWhiteSpace = Pattern.compile("(\\s*(\\d)\\s*([+\\-*/])\\s*)+(\\d)\\s*")
```
ExpressionValidator가 생성될때마다 고비용인 Pattenr객체가 계속해서 생성될 수 있을것 같습니다!

ExpressionValidator가 patternThatAllowWhiteSpace이외 별다른 상태값을 가지고 있지 않으니 class가 아닌 object 객체로 만들어보는것도 좋을것 같습니다!

pci2676 replied:
혹시 주어진 요구사항중에 입력받는 숫자에 대한 제한이 있었을까요?

10 + 10을 하면 문제가 발생할 것 같아보여서요~!

maenguin replied:
(피드백 반영 순서를 기입해두면 코드 변경 히스토리 추적에 용이하실 것 같아서 번호를 앞에 표시하겠습니다.)


[1] 커밋 [ce821d3](https://github.com/next-step/kotlin-racingcar/pull/1072/commits/ce821d34a87c27e659e8a2b72880beec1381c4cb) 으로 피드백 반영했습니다.

> ExpressionValidator가 생성될때마다 고비용인 Pattenr객체가 계속해서 생성될 수 있을것 같습니다!
> 
> ExpressionValidator가 patternThatAllowWhiteSpace이외 별다른 상태값을 가지고 있지 않으니 class가 아닌 object 객체로 만들어보는것도 좋을것 같습니다!

Pattern 객체를 static으로 처리했습니다.
ExpressionValidator의 현재 상태와 비용을 고려했을때 object로 바꾸는게 더 적합한 상황 같네요 참고하겠습니다!

> 혹시 주어진 요구사항중에 입력받는 숫자에 대한 제한이 있었을까요?
> 
> 10 + 10을 하면 문제가 발생할 것 같아보여서요~!

두 자리 이상의 숫자는 처리가 불가능한 상태였네요; 두 자리 이상 처리 가능하도록 변경 후 테스트 반영했습니다.

pci2676 commented on src/main/kotlin/step2/ExpressionValidator.kt:
```diff
@@ -0,0 +1,19 @@
+    private val patternThatAllowWhiteSpace = Pattern.compile("(\\s*(\\d)\\s*([+\\-*/])\\s*)+(\\d)\\s*")
+
+    fun validate(expression: String?) {
+        require(!expression.isNullOrEmpty()) { "Expression must not be a null or empty" }
```
꼼꼼하게 체크해 주셨네요! 👍

pci2676 commented on src/test/kotlin/step2/ExpressionValidatorTest.kt:
```diff
@@ -0,0 +1,51 @@
+        val sut = ExpressionValidator()
+        assertThatThrownBy { sut.validate("") }
+            .isInstanceOf(IllegalArgumentException::class.java)
+    }
```
`@NullAndEmptySource`를 이용해서 하나의 테스트로 통합할 수 있을것 같아보이네요!

maenguin replied:
[2] 커밋 [f8ee373](https://github.com/next-step/kotlin-racingcar/pull/1072/commits/f8ee3735ed64b88007fc4e0381b05e7fdd241adc) 으로 반영했습니다.
이런 애노테이션도 있었군요 배워갑니다 👍

pci2676 commented on src/main/kotlin/step2/Operand.kt:
```diff
@@ -0,0 +1,46 @@
+
+    operator fun div(target: Operand): Operand {
+        return Operand((value / target.value).toString())
+    }
```
연산자 오버라이드를 적절하게 적용하셨군요 👍 
관련해서 이펙티브 코틀린에도 내용이 있으니 한번 읽어보시는걸 추천드립니다 😄 

아이템 12 - 연산자 오버로드를 할 때는 의미에 맞게 사용하라

pci2676 commented on src/main/kotlin/step2/Operand.kt:
```diff
@@ -0,0 +1,46 @@
+
+    override fun hashCode(): Int {
+        return value.hashCode()
+    }
```
`Operand`도 data 클래스로 만들었으면 이와 같은 메서드는 구현을 코틀린에게 맡길 수 있을것 같은데요!
data 클래스로 하지 않은 이유가 있을까요?

maenguin replied:
[3] 커밋 [8163d19](https://github.com/next-step/kotlin-racingcar/pull/1072/commits/8163d193d2fb2a077e0a01b735b84d73c71c66a9)으로 반영했습니다.
그러게요; data 클래스로 했어도 무방했네요 제가 잠시 착각했었나봅니다.

pci2676 commented on src/main/kotlin/step2/CalculatorStrategy.kt:
```diff
@@ -0,0 +1,18 @@
+        }
+        return operand.toString()
+    }
+}
```
`expressionItems: List<ExpressionItem>`를 멤버변수로 갖는 객체에게 이 책임을 위임해보는건 어떨까요?

maenguin replied:
CalculatorStrategy가 expressionItems: List<ExpressionItem>를 멤버변수로 갖게 처리해도 해답이 될 수 있을까요?

pci2676 replied:
그렇게해도 나쁘지 않을것 같습니다~!

pci2676 commented on src/main/kotlin/step2/Main.kt:
```diff
@@ -0,0 +1,12 @@
+        expressionValidator = ExpressionValidator(),
+        expressionTokenizer = ExpressionTokenizer(),
+        expressionParser = ExpressionParser(),
+        calculatorStrategy = CalculatorStrategy()
```
이름있는 아규먼트를 잘 사용해 주셨네요! 👍

pci2676 commented on src/test/kotlin/step2/CalculatorTest.kt:
```diff
@@ -0,0 +1,28 @@
+        "1+2, 3",
+        "2 * 2 * 2, 8"
+    )
+    internal fun 성공(expression: String, expected: String) {
```
메서드 이름에 한글을 적는것도 나쁘지 않지만
`@DisplayName`을 사용하는 것도 고려해보시면 좋을 것 같습니다!

maenguin replied:
저도 원래 @DisplayName를 사용하는데 영어 메서드명 짓는게 고민이 많이 되더라구요
그래서 현재 처럼 한글 메서드 사용하는게 편한거 같습니다.

pci2676 commented on src/main/kotlin/step2/strategy/CalculatorStrategy.kt:
```diff
@@ -0,0 +1,22 @@
+ */
+class CalculatorStrategy {
+
+    fun apply(expressionItems: List<ExpressionItem>): String {
+        var operand = expressionItems.first() as Operand
+        for (i in 1 until expressionItems.size step 2) {
+            val operator = expressionItems[i] as Operator
+            operand = operator.operate(operand, expressionItems[i + 1] as Operand)
+        }
+        return operand.toString()
```
입력받은 `expressionItems`의 상태에 따라 메서드에서 예외 상황이 많이 발생할 것 같습니다!

`expressionItems`인 List의 여러가지 메서드를 사용하고 있는데 일급컬렉션으로 추상화 해보는건 어떨까요?

[일급컬렉션](https://jojoldu.tistory.com/412)

maenguin replied:
말씀주신대로 ExpressionItems 라는 일급컬렉션으로 추상화시켜보려고 합니다.
ExpressionItems는 문자 연산식으로 부터 추출한 완벽히 검증된 연산 요소를 담은 자료구조가 될 것 같습니다.

요구사항에는 없지만 계산기의 연산 전략(우선순위 고려 하지 않음)이 런타임에서 바뀔 수 있다고 가정할때
현재처럼 CalculatorStrategy를 사용하려면 ExpressionItems이 CalculatorStrategy를 의존하는 형태로 바꿔야할까요?

pci2676 replied:
전략은 유지하되 List에 담긴 요소를 사용하려고 한다면 CalculatorStrategy가 ExpressionItems 일급컬렉션을 의존하는방향이 좋을것 같다는 생각입니다.
ExpressionItems이 자신을 계산하는 전략에 의존한다면 전략이 늘어날때마다 계속해서 의존성이 늘어나게 될 것 같습니다.

pci2676 commented on src/main/kotlin/step2/Calculator.kt:
```diff
@@ -0,0 +1,25 @@
+) {
+
+    fun calculate(expression: String?): String {
+        expressionValidator.validate(expression)
+        val expressionTokens: List<String> = expressionTokenizer.tokenize(expression!!)
```
validate와 tokenize는 반드시 순서가 보장되어야 올바른 결과 값을 얻을 수 있을것 같아보이는데요.

`expression!!`으로 값을 확신하는 것보다 validator의 동작을 tokenizer에서 수행해보는걸 고려해보셔도 좋을것 같습니다.

두 객체가 입력받는 매개변수가 동일해서 변경사항이 그리 많지 않을것 같아 보입니다!

maenguin replied:
[5] 커밋 [0003817](https://github.com/next-step/kotlin-racingcar/pull/1072/commits/00038174379392673fc479685af08a06fdcc7610)으로 반영했습니다.

검증된 연산식으로 tokenize를 수행해야되기 때문에 tokenizer에서 validate를 수행하는게 응집도 있고 안전한 코드가 될것 같아서
말씀주신대로 반영했습니다.

pci2676 commented on src/main/kotlin/step2/parser/ExpressionParser.kt:
```diff
@@ -0,0 +1,19 @@
+class ExpressionParser {
+    fun parse(expressionTokens: List<String>): List<ExpressionItem> {
+        return expressionTokens.map {
+            when (it.toDoubleOrNull()) {
```
token으로 전달받은 String이 스스로 double인지 아닌지를 판단해서 Operator로 변화하고 있는것으로 보이는데요.  


Parser가 아닌 token 스스로 자신의 상태를 판단해서 Operator로 변화하는건 어떨까요?

maenguin replied:
[4] 커밋 [6d451f0](https://github.com/next-step/kotlin-racingcar/pull/1072/commits/6d451f0bbd3d9561c69124f0327714ab0825a655)으로 반영했습니다.

ExpreesionToken을 추가하고 ExpreesionToken이 자신의 상태를 판단해서 ExpressionItem으로 변환하도록 했습니다.

pci2676 commented on src/main/kotlin/step2/strategy/CalculatorStrategy.kt:
```diff
@@ -0,0 +1,22 @@
+class CalculatorStrategy {
+
+    fun apply(expressionItems: List<ExpressionItem>): String {
+        var operand = expressionItems.first() as Operand
+        for (i in 1 until expressionItems.size step 2) {
+            val operator = expressionItems[i] as Operator
+            operand = operator.operate(operand, expressionItems[i + 1] as Operand)
```
인터페이스로 입력받았는데  다운캐스팅 하면서 사용해야 한다면 

인터페이스로 추상화하는게 반드시 필요했는지? 
아니면 다운캐스팅을 하지 않고 문제를 해결할 수 있는 방법이 있는지?

의심해보면 좋을것 같습니다. 🤨

maenguin replied:
연산자 타입과 피연산자 타입을 개념적으로 묶는게 용이하다고 봐서 인터페이스로 묶었는데요
너무 과도한 조치였을까요?

pci2676 replied:
`as`를 이용하기 위한 최소한의 안정장치라도 ExpressionItem이 제공해주는것 어떨까요?
abstract val isOperator: Boolean 을 제공하였다면 ExpressionItem을 사용하는 입장에서 조금 더 안전하게 사용할 수 있지 않았을까? 라는 생각이 듭니다.

지금은 순서에만 의존하여 변경하고 있으니까요!

pci2676 reviewed (changes requested):
전체적으로 간결하게 작성해주시고 커밋을 세세하게 쪼개주신덕에 코드를 파악하기 아주 수월했던것 같습니다 👍 

조금 더 생각해보면 좋을점과 개선해보면 좋을 것 같다고 생각하는 부분에 코멘트를 남겨두었습니다.
반드시 모든 부분을 반영하실 필요는 없고 곰곰히 생각해보시고 반영해주시면 될 것 같습니다!

한번 더 힘내서 리뷰요청 부탁드리겠습니다! 🙂

maenguin committed:
step2 : 피드백 반영 - Pattern 객체 생성 비용, 숫자 자릿 수 문제

- Pattern 객체 한 번만 생성되도록 변경
- 숫자 자릿 수 두 자리 이상 가능하도록 변경 및 관련 테스트 추가

maenguin committed:
step2 : 피드백 반영 - @NullAndEmptySource 적용

maenguin committed:
step2 : 피드백 반영 - Operand class에 data 클래스 적용

maenguin committed:
step2 : 피드백 반영 - Token의 책임

- 연산식에서 추출한 토큰을 의미하는 ExpressionToken 객체 추가 및 ExpressionItem으로 변환 책임 수행
- 관련 테스트 수정

maenguin committed:
step2 : 피드백 반영 - Tokenizer와 Validator 통합

- Tokenizer에서 검증수행하도록 변경
- Validator를 object 클래스로 변경
- Validator를 tokenzier 패키지로 이동

pci2676 reviewed (approved):
코드리뷰 반영해주신것 확인했습니다!

충분히 깔끔하게 작성해주셔서 이번 미션은 여기서 마무리짓고 본 미션으로 넘어가는게 좋을것 같습니다!

수고하셨습니다!
